tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py

```diff
@@ -229,19 +249,80 @@ def test_evaluate_cfrag(testerchain, escrow, adjudicator_contract):
-    assert value - BASE_PENALTY == escrow.functions.minerInfo(miner).call()[0]
-    assert BASE_PENALTY / REWARD_COEFFICIENT == escrow.functions.rewardInfo(investigator).call()
+
+    penalty, reward = compute_penalty_and_reward(worker_stake, worker_penalty_history)
+    worker_stake -= penalty
+    investigator_balance += reward
+    worker_penalty_history += 1
+
+    assert worker_stake == escrow.functions.minerInfo(miner).call()[0]
+    assert investigator_balance == escrow.functions.rewardInfo(investigator).call()
 
     events = evaluation_log.get_all_entries()
-    assert 2 == len(events)
-    event_args = events[1]['args']
+    assert number_of_evaluations == len(events)
+    event_args = events[-1]['args']
     assert data_hash == event_args['evaluationHash']
     assert miner == event_args['miner']
     assert investigator == event_args['investigator']
     assert not event_args['correctness']
 
+    ###############################
+    # Test: Bob produces wrong precomputed data
+    ###############################
+
+    metadata = os.urandom(34)
+    capsule, cfrag = fragments(metadata)
+    capsule_bytes = capsule.to_bytes()
+    cfrag_bytes = cfrag.to_bytes()
+    assert cfrag.verify_correctness(capsule)
+
+    hash_ctx = hashes.Hash(hashes.SHA256(), backend=backend)
+    hash_ctx.update(capsule_bytes + cfrag_bytes)
+    data_hash = hash_ctx.finalize()
+    capsule_signature_by_requester = sign_data(capsule_bytes, requester_umbral_private_key)
+    capsule_signature_by_requester_and_miner = sign_data(capsule_signature_by_requester, miner_umbral_private_key)
+    cfrag_signature_by_miner = sign_data(cfrag_bytes, miner_umbral_private_key)
+    evidence = IndisputableEvidence(capsule, cfrag, ursula=None)
+    evidence_data = evidence.precompute_values()
+
+    # Bob produces a random point and gets the bytes of coords x and y
+    random_point_bytes = Point.gen_rand().to_bytes(is_compressed=False)[1:]
+    # He uses this garbage instead of correct precomputation of z*E
+    evidence_data = bytearray(evidence_data)
+    evidence_data[32:32+64] = random_point_bytes
+    evidence_data = bytes(evidence_data)
+
+    args = (capsule_bytes,
+            capsule_signature_by_requester,
+            capsule_signature_by_requester_and_miner,
+            cfrag_bytes,
+            cfrag_signature_by_miner,
+            requester_umbral_public_key_bytes,
+            miner_umbral_public_key_bytes,
+            signed_miner_umbral_public_key,
+            evidence_data)
+
+    assert not adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
+
+    # Evaluation must fail since Bob precomputed wrong values
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = adjudicator_contract.functions.evaluateCFrag(*args).transact({'from': investigator})
+        testerchain.wait_for_receipt(tx)
+
+    ###############################
+    # Test: Signatures and Metadata mismatch
+    ###############################
+
+    # TODO
```

Comment:
Is this going to be in another PR?

Reply:
Yes, additional work is needed before we can test anything.

Reply:
Can you link the newly created issues in the PR, please?

Reply:
#626